File: problems/003/003.go
package main

import "fmt"
import "os"
import "strconv"
import "time"

func f() uint64 {
    n, factor := uint64(600851475143), uint64(2)

    for {
        if n % factor == 0 {
            n /= factor
        } else {
            factor += 1
        }

        if factor * factor > n {
            return n
        } else if n == 1 {
            return factor
        }
    }
}

func main() {
    if len(os.Args) == 1 {
        fmt.Println(f())
    } else if len(os.Args) == 2 {
        iters, _ := strconv.ParseInt(os.Args[1], 10, 64)

        start := time.Now()
        for i := int64(0); i < iters; i++ {
            f()
        }
        end := time.Now()

        fmt.Println(end.Sub(start).Nanoseconds())
    }
}

File: problems/003/003.rs
extern crate test;
extern crate time;

use std::os::args;
use test::black_box;
use time::precise_time_ns;

#[inline]
fn f() -> u64 {
    let (mut n, mut factor) = (600851475143, 2);

    loop {
        if n % factor == 0 {
            n /= factor;
        } else {
            factor += 1;
        }

        if factor * factor > n {
            return n;
        } else if n == 1 {
            return factor;
        }
    }
}

fn main() {
    match args().as_slice() {
        [_] => {
            println!("{}", f());
        },
        [_, ref iters] => {
            let iters: u64 = from_str(iters.as_slice()).unwrap();

            let start = precise_time_ns();
            for _ in range(0, iters) {
                black_box(f());
            }
            let end = precise_time_ns();

            println!("{}", end - start);
        },
        _ => unreachable!(),
    }
}
